cpp/src/dataframe/device_table.cuh

```diff
@@ -0,0 +1,543 @@
+   * 
+   * @return The hash value of the row
+   * ----------------------------------------------------------------------------**/
+  template <template <typename> class hash_function = default_hash>
+  __device__ 
+  hash_value_type hash_row(gdf_size_type row_index,
+                           hash_value_type* initial_hash_values = nullptr,
+                           gdf_size_type num_columns_to_hash = 0) const
+  {
+    hash_value_type hash_value{0};
+
+    // If num_columns_to_hash is zero, hash all columns
+    if(0 == num_columns_to_hash) 
+    {
+      num_columns_to_hash = this->_num_columns;
+    }
+
+    bool const use_initial_value{ initial_hash_values != nullptr };
+    // Iterate all the columns and hash each element, combining the hash values together
+    for (gdf_size_type i = 0; i < num_columns_to_hash; ++i) {
+
+      // Skip null values
+      if (gdf_is_valid(d_columns_valids_ptr[i], row_index)) {
+        gdf_dtype const current_column_type = d_columns_types_ptr[i];
+        hash_value_type const initial_hash_value =
+            (use_initial_value) ? initial_hash_values[i] : 0;
+        cudf::type_dispatcher(current_column_type,
+                              hash_element<hash_function>{}, hash_value,
+                              d_columns_data_ptr[i], row_index, i,
+                              use_initial_value, initial_hash_value);
+      }
+    }
+
+    return hash_value;
+  }
+
+private:
+
+  const gdf_size_type _num_columns; /** The number of columns in the table */
+  gdf_size_type _num_rows{0};     /** The number of rows in the table */
+
+  gdf_column ** host_columns{nullptr};  /** The set of gdf_columns that this table wraps */
+
+  rmm::device_vector<void*> device_columns_data; /** Device array of pointers to each columns data */
+  void ** d_columns_data_ptr{nullptr};                  /** Raw pointer to the device array's data */
+
+  rmm::device_vector<gdf_valid_type*> device_columns_valids;  /** Device array of pointers to each columns validity bitmask*/
+  gdf_valid_type** d_columns_valids_ptr{nullptr};                   /** Raw pointer to the device array's data */
+
+  rmm::device_vector<gdf_valid_type> device_row_valid;  /** Device array of bitmask for the validity of each row. */
+  gdf_valid_type * d_row_valid{nullptr};                   /** Raw pointer to device array's data */
+
+  rmm::device_vector<gdf_dtype> device_columns_types; /** Device array of each columns data type */
+  gdf_dtype * d_columns_types_ptr{nullptr};                 /** Raw pointer to the device array's data */
+
+  gdf_size_type row_size_bytes{0};
+  rmm::device_vector<gdf_size_type> device_column_byte_widths;
+  gdf_size_type * d_columns_byte_widths_ptr{nullptr};
+
+protected:
+ /**---------------------------------------------------------------------------*
+  * @brief Constructs a new device_table object from an array of `gdf_column*`.
+  *
+  * This constructor is protected to require use of the device_table::create
```

Comment:
Yeah, I've definitely thought about separating out that step to be optional. I just haven't figured out what to do with the `row_has_nulls` function if the `device_table` was created without generating the ANDed bitmask.

Reply:
Okay, I figured out what to do to avoid always computing the row-level bitmask. I definitely want to maximize utility of this class, and you're totally right that it's too heavyweight to always compute that. It's just not good class design either.

Reply:
See the vastly updated implementation.

`device_table` is now a trivially lightweight wrapper for an array of `gdf_column`s in device memory. It provides minimal convenience accessor member functions. All other functionality was moved to free functions. Furthermore, it no longer uses managed memory.